### DISCUSSION TITLE
Trouble with redirecting in next.config

### DISCUSSION THREAD

**Discussion by bscaspar** [vercel/next.js#52611] [Help] [answered]
### Summary

I am trying to create some _slightly_ complicated redirects in the `next.config` file and am wondering if there's a simpler method. We're doing some consolidation/refactoring and some of our pages with catch-all and optional catch-all params will be consolidated into one place. Here's an example:

1. Redirect for `/a/[[...slug]]`, here are the different desired results:
  - `/a  ->  /`
  - `/a/abc  ->  /?param=abc`
  - `/a/abc/xyz  ->  /?param=abc` (dropping the `xyz` param and any others after it)

My solution: 
```
async redirects() {
  return [
    {
      source: '/a',
      destination: '/',
      permanent: false,
    },
    {
        source: '/a/:slug',
        destination: '/?param=:slug',
        permanent: false,
    },
    {
       source: '/a/:slug/:otherSlug*',
       destination: '/?param=slug',
       permanent: false,
     },
  ]
}
```
I was hoping to use some sort of optional catch-all pattern for `source` but ran into errors converting the catch-all slug into the query string.

Is this the simplest way to accomplish the goals I have?

**Comment by rasif-sahl** [comment]
Your approach of using separate redirect rules in the next.config.js file to achieve the desired results is a valid approach. However, there is a simpler way to accomplish the goals you have by using a single rewrite rule with a custom rewrite function.

Here's an example of how you can simplify the redirects:

```
// next.config.js

async function rewriteSlug(req, res, options) {
  const { slug } = req.query;
  let param = '';

  if (Array.isArray(slug)) {
    param = slug[0];
  } else if (slug) {
    param = slug;
  }

  res.writeHead(307, {
    Location: `/?param=${param}`,
  });
  res.end();
}

module.exports = {
  async rewrites() {
    return [
      {
        source: '/a/:slug*',
        destination: '/api/rewrite-slug',
      },
    ];
  },
  async api() {
    return {
      externalResolver: true,
      bodyParser: false,
      routes: async () => [
        {
          source: '/api/rewrite-slug',
          destination: '/api/rewrite-slug',
          rewrite: rewriteSlug,
        },
      ],
    };
  },
};
```
In this example, a single rewrite rule is used with the source pattern /a/:slug*, which matches all the variations of the /a route with optional slugs. The destination points to an API route /api/rewrite-slug.

The rewriteSlug function is defined as a custom rewrite function, which is called for the /api/rewrite-slug API route. It extracts the slug from the request query parameters and constructs the desired destination URL using a query string. The 307 status code indicates a temporary redirect.

By using this approach, you can simplify the redirects and achieve the desired results with a single rewrite rule and a custom rewrite function.

Remember to adjust the destination URLs, status codes, and any other parameters as per your specific requirements.

**Reply by bscaspar** [reply]
Thanks for the response @rasif-sahl! This seems like a great option, moving the logic to an API route is a good idea, especially as it gets more complicated. I personally think I will implement something similar but in the individual page files, or in middleware, as @icyJoseph suggested, to keep the `next.config` file a little simpler.

**Comment by icyJoseph** [comment, marked as answer]
Under the hood, IIRC, https://github.com/pillarjs/path-to-regexp is used here, so you could simplify your rewrites a bit futher:

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  async redirects() {
    return [
      {
        source: "/a",
        destination: "/",
        permanent: false,
      },
      {
        source: "/a/:slug/:bar?",
        destination: "/?param=:slug",
        permanent: false,
      },
    ];
  },
};

module.exports = nextConfig;
```

Though, I wonder if middleware or an actual `getServerSideProps` in `/a/[[...slug]].tsx` are, perhaps, better options?

Both middleware and `getServerSideProps` for `/a/[[...slug]].tsx`, could look at the incoming request, take the `slug` parameter, and redirect as needed.

```ts
// src/pages/a/[[...slug]].tsx
export const getServerSideProps = (ctx: GetServerSidePropsContext) => {
  const slug = ctx.params?.slug;

  if (!Array.isArray(slug)) {
    return { redirect: { destination: "/", permanent: false } };
  }

  const destination = slug ? `/?param=${slug[0]}` : "/";

  return {
    redirect: {
      destination,
      permanent: false,
    },
  };
};

```

Middleware would be something similar, but checking `req.nextUrl.pathname`, and redirecting based off that.

**Reply by bscaspar** [reply]
Thanks @icyJoseph, that simplification worked well. 

Your point about moving to custom server logic, instead of the config, similar to @rasif-sahl's suggestion, makes a lot of sense. Especially since some of the redirects we are making are a little more complicated than this example, and some of the logic currently relies on the specific order of each redirect in the `next.config` file - probably best to avoid that for maintainability.
